flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/ZeroCopyNettyMessageDecoder.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandlerAdapter;
+
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.FRAME_HEADER_LENGTH;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.MAGIC_NUMBER;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Decodes messages from the fragmentary netty buffers. This decoder assumes the
+ * messages have the following format:
+ * +-----------------------------------+--------------------------------+
+ * | FRAME_HEADER ||  MESSAGE_HEADER   |     DATA BUFFER (Optional)     |
+ * +-----------------------------------+--------------------------------+
+ * and it decodes each part in order.
+ *
+ * This decoder tries best to eliminate copying. For the frame header and message header,
+ * it only cumulates data when they span multiple input buffers. For the buffer part, it
+ * copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class ZeroCopyNettyMessageDecoder extends ChannelInboundHandlerAdapter {
+
+    private static final int INITIAL_MESSAGE_HEADER_BUFFER_LENGTH = 128;
+
+    /** The parser to parse the message header. */
+    private final NettyMessageParser messageParser;
+
+    /** The buffer used to cumulate the frame header part. */
+    private ByteBuf frameHeaderBuffer;
+
+    /** The buffer used to receive the message header part. */
+    private ByteBuf messageHeaderBuffer;
+
+    /** Which part of the current message is being decoded. */
+    private DecodeStep decodeStep;
+
+    /** How many bytes have been decoded in current step. */
+    private int decodedBytesOfCurrentStep;
```

Comment:
The semantic of this var is hard to understand. Actually it is only used for judging whether the full `BufferResponse` is decoded in the third step and then fired to the next handler. So it is useless for the first and second steps, but we still set it to 0 in these two steps which seem reductant and ambiguous.

Reply:
Agree with that the variable is only meaningful for decoding the Buffer part of the message. However, I still think that it could not be removed, since we need to bookkeeping how many bytes have been discarded already.